### ISSUE TITLE
Bughunting: regress-store-holey-double-array

### ISSUE THREAD

**Issue by palfia** [v8mips/v8mips#337]
Fails on MIPS64 SIM since: https://codereview.chromium.org/2060233002

**Comment by ilijapavlovic** [comment]
The test has defined an array with a hole: `a = [,0.1]`.

The first execution of the test on simulator is without `--turbo` option:

```
sim> printobject 0x22c11819
0x22c11819: 
0x22c11819: [JSArray]
 - map = 0x3e806bf1 [FastProperties]
 - prototype = 0x7f1bb79a9e61
 - elements = 0x22c11849 <FixedDoubleArray[2]> [FAST_HOLEY_DOUBLE_ELEMENTS]
 - length = 2
 - properties = {
   #length: 0x7f1bb7981ee9 <AccessorInfo> (accessor constant)
 }
 - elements = {
   0: <the hole>
   1: 0.1
 }

sim> disasm
  0x3b71c384   463c0781       sub.d   f30, f0, f28    // 0x7ff8000000000000
  0x3b71c388   f67e0000       sdc1    f30, 0(s3)      // 0x7ff8000000000000 --> [0x22c11858] ; stored in array
  0x3b71c38c   00801025       mov     v0, a0          // 0x22c119b1
  0x3b71c390   03e00008       jr      ra              // --> 0x3b795264

sim> printobject 0x22c11819
0x22c11819: 
0x22c11819: [JSArray]
 - map = 0x3e806bf1 [FastProperties]
 - prototype = 0x7f1bb79a9e61
 - elements = 0x22c11849 <FixedDoubleArray[2]> [FAST_HOLEY_DOUBLE_ELEMENTS]
 - length = 2
 - properties = {
   #length: 0x7f1bb7981ee9 <AccessorInfo> (accessor constant)
 }
 - elements = {
   0: nan
   1: 0.1
 }
```

The second test execution is with `--turbo` option  (problematical execution):

```
sim> printobject 0x22c11819
0x22c11819: 
0x22c11819: [JSArray]
 - map = 0x3e806bf1 [FastProperties]
 - prototype = 0x7fe7f1ba9e61
 - elements = 0x22c11849 <FixedDoubleArray[2]> [FAST_HOLEY_DOUBLE_ELEMENTS]
 - length = 2
 - properties = {
   #length: 0x7fe7f1b81ee9 <AccessorInfo> (accessor constant)
 }
 - elements = {
   0: <the hole>
   1: 0.1
 }

sim> disasm
  ...
  0x3b795b78   463c0001       sub.d   f0, f0, f28  // 0xffff7fffffff7fff (-nan)
  0x3b795b7c   00c4102d       daddu   v0, a2, a0   // 0x22c11858
  0x3b795b80   f4400000       sdc1    f0, 0(v0)    // 0xffff7fffffff7fff (-nan) --> [0x22c11858]
  ...

sim> printobject 0x22c11819
0x22c11819: 
0x22c11819: [JSArray]
 - map = 0x3e806bf1 [FastProperties]
 - prototype = 0x7fe7f1ba9e61
 - elements = 0x22c11849 <FixedDoubleArray[2]> [FAST_HOLEY_DOUBLE_ELEMENTS]
 - length = 2
 - properties = {
   #length: 0x7fe7f1b81ee9 <AccessorInfo> (accessor constant)
 }
 - elements = {
   0: <the hole>
   1: 0.1
 }
```

The array element `a[0]` is stored on memory location `0x22c11858`. Differences between these tests are following:
- In the first case, stored value is NaN (`0x7ff8000000000000`). In the second case, stored value is -NaN (`0xffff7fffffff7fff`). Mentioned values are stored on the same location `0x22c11858` (element `a[0]`).
- In memory location `0x22c11860` (element `a[1]`) is stored value `0x3fb999999999999a` which represents floating-point format for decimal `0.1`.
- For the second version of the test, the strange thing is output for `printobject 0x22c11819`. Although -NaN value is stored in address for `a[0]` element (`0x22c11858`) it is not shown in output: `0: <the hole>` instead `0: -nan`.

**Comment by ilijapavlovic** [comment]
Also, the interesting thing here is that stored values `0x7ff8000000000000` or `0xffff7fffffff7fff` are actually signalling NaNs - not quiet NaNs. (CL title: _"[turbofan] Prevent storing signalling NaNs into holey double arrays."_)

**Comment by ilijapavlovic** [comment]
Checked execution on Qemu. NaNs are represented as quiet NaNs.

```
(gdb) x /5vi $pc-4
   0x54d1c380:  daddu   s3,s3,at
=> 0x54d1c384:  sub.d   $f30,$f0,$f28
   0x54d1c388:  sdc1    $f30,0(s3)
   0x54d1c38c:  move    v0,a0
   0x54d1c390:  jr  ra
   0x54d1c394:  nop

(gdb) info registers f30 f0 f28
f30: 0x7ff7ffffffffffff flt: -nan              dbl: nan                     
f0:  0x7ff7ffffffffffff flt: -nan              dbl: nan                     
f28: 0x0000000000000000 flt: 0                 dbl: 0 
```

**Comment by ilijapavlovic** [comment]
The test will pass on simulator if following conditions are fulfilled for NaN:
- bit 63: set on 0 (cleared)
- bit 51: set on 1

Condition that the bit 51 must be set on 1 can be found in `./src/./src/objects-debug.cc:`

```
void FixedDoubleArray::FixedDoubleArrayVerify() {
  for (int i = 0; i < length(); i++) {
    if (!is_the_hole(i)) {
      uint64_t value = get_representation(i);
      uint64_t unexpected =
          bit_cast<uint64_t>(std::numeric_limits<double>::quiet_NaN()) &
          V8_UINT64_C(0x7FF8000000000000);

      // Create implementation specific sNaN by inverting relevant bit.
      unexpected ^= V8_UINT64_C(0x0008000000000000);

      CHECK((value & V8_UINT64_C(0x7FF8000000000000)) != unexpected ||
            (value & V8_UINT64_C(0x0007FFFFFFFFFFFF)) == V8_UINT64_C(0));
    }
  }
}
```

Simplified: `(value & 0x7ff8000000000000) != 0x7ff0000000000000`
- This means that in NaN the bit 51 must be set.

When test performs checking `IsNaN()` it seems from the simulator trace that it is done against `0x7ff8000000000000`.

Mentioned requirements can be implemented with changes on `simulator-mips64.cc`:

```
    case SUB_D: {
      double res = fs - ft;
      if (std::isnan(res)) {
        *reinterpret_cast<uint64_t*>(&res) |= (1L << 51);
        *reinterpret_cast<uint64_t*>(&res) &= ~(1L << 63);
      }
      set_fpu_register_double(fd_reg(), res);
      TraceRegWr(FPUregisters_[fd_reg()]);
      break;
    }
```

This is because of problematical CL which is implemented on following way:

```
src/compiler/mips64/code-generator-mips64.cc:

case kMips64Float64SilenceNaN:
  __ FPUCanonicalizeNaN(i.OutputDoubleRegister(), i.InputDoubleRegister(0));
  break;

-----------------------------
./src/mips64/macro-assembler-mips64.cc:

void MacroAssembler::FPUCanonicalizeNaN(const DoubleRegister dst,
                                        const DoubleRegister src) {
  sub_d(dst, src, kDoubleRegZero);
}
```

Therefore, adaptation SUB_D for simulator is selected as most suitable place for adaptation.

**Comment by kilvadyb** [comment]
Nice findings. I don't think we should patch the simulator to pass the test. Have you tried it on boards and qemur6? At the first look I think the `-nan <-> nan` should be fixed in turbofan code generation. On simulator r2 and r6 we have to use intel nans which are the same as on MIPSr6. We use different nans on r1 and r2 boards only.

**Comment by ilijapavlovic** [comment]
The test is performed for mips64el and here are results of testing:
- Board r2: successful test.
- Qemu r2: successful test.
- Qemu r6: the test fails (all three option combinations of the test are failed).
- Simulator r2: the test fails with `--turbo` option.
- Simulator r6: successful test.

After these results, turbofan code generation really should be investigated for NaNs.

**Comment by kilvadyb** [comment]
The last 2 results (of sims) is a bit surprising as i387 intel (and r6) nans should be used in both cases I think so both should fail or succeed.

**Comment by ilijapavlovic** [comment]
CL uploaded and ready for review: https://codereview.chromium.org/2137163003/

**Comment by kilvadyb** [comment]
At the last call of `MacroAssembler::FPUCanonicalizeNaN()` the f0 has different values on r2 and r6 simulators:

```
r6:
f0: 0xfff7fffffff7ffff

r2:
f0: 0xffff7fffffff7fff
```

**Comment by ilijapavlovic** [comment]
`code-generator-mips64.cc` has implementation as

```
    case kMips64Float64SilenceNaN: {
      __ FPUCanonicalizeNaN(i.OutputDoubleRegister(), i.InputDoubleRegister(0));
      break;
    }
```

This is tested on FPGA r6 board and the test is executed successfully. This means, that qemu mips64el in user mode has a problem if mentioned implementation is compiled for r6.

**Comment by kilvadyb** [comment]
1st proposal CL: https://codereview.chromium.org/2171303002/

**Comment by kilvadyb** [comment]
Fix of the proposal/reverted CL: https://codereview.chromium.org/2188433002/

**Comment by kilvadyb** [comment]
I am also working on to add mips-specific NaN handling to simulator. Could be a new issue? Proposal CL: https://codereview.chromium.org/2184843002/

**Comment by kilvadyb** [comment]
Fix CL landed, new issue (#367) is opened for simulator improvement.
